features/features_test.go

```diff
@@ -378,59 +378,92 @@ func (s *featureSuite) TestAll(c *C) {
 	defer st.Unlock()
 	tr := config.NewTransaction(st)
 
+	// Use the first feature as a testing feature and overwrite all its values.
+	fakeFeature := features.SnapdFeature(0)
```

Comment:
`SnapdFeature(0)` is actually already defined as `Layouts`. I'd prefer to use something like this, as we did for `TestIsSupported`:
```suggestion
	fakeFeature := features.SnapdFeature(len(features.KnownFeatures()))
```

Reply:
There might be a problem though, because `All` calls `KnownFeatures`, which depends on `lastFeature` (which will not include `SnapdFeature(len(features.KnownFeatures()))`, so maybe it is better to do what you're already doing.

Reply:
Another piece may be to mock `KnownFeatures` and/or `FeaturesSupportedCallbacks` entirely,  though I'm not sure whether it would actually be simpler to do it this way. Then there wouldn't be complexity around restoring values in the original, though?

Reply:
I was concerned about mocking `KnownFeatures`, which is why I chose this other path. My reasoning could be entirely flawed though especially since I really know very little about golang.

`KnownFeatures` has the following properties:
- exported
- when called, will always return the same list of features

To mock `KnownFeatures`, I would need to transform it from a function into a variable. To have it maintain its same properties, I would need to protect that variable from outside mutations. So I would create an internal variable that contains the array of features. The exported function then would always make a copy of that array so as to protect against outside interference. 

It seemed to me that the modifications would slightly increase the code complexity of features.go, and I thought it better to instead slightly increase the complexity on the test. Again, though, I'm very new to golang-style writing so I have no doubt my thought process is off base. Would mocking `KnownFeatures` be more in line with golang and snapd style @olivercalder?

Reply:
Go is also relatively new to me, so I'm not sure if one way is certainly better than another. I do know we have some tricks to allow us to essentially mock exported functions without actually mocking them, with minimal code complexity. Something like this:
```
diff --git a/features/export_test.go b/features/export_test.go
index dad572b50d..e24d003b28 100644
--- a/features/export_test.go
+++ b/features/export_test.go
@@ -37,3 +37,7 @@ func MockReleaseSystemctlSupportsUserUnits(f func() bool) (restore func()) {
        releaseSystemctlSupportsUserUnits = f
        return r
 }
+
+func MockKnownFeaturesImpl(f func() []features.SnapdFeature) (restore func()) {
+       return testutil.Mock(&knownFeaturesImpl, f)
+}
diff --git a/features/features.go b/features/features.go
index aa916b5ecf..40fd709b47 100644
--- a/features/features.go
+++ b/features/features.go
@@ -82,8 +82,7 @@ const (
        lastFeature
 )
 
-// KnownFeatures returns the list of all known features.
-func KnownFeatures() []SnapdFeature {
+var knownFeaturesImpl = func() []SnapdFeature {
        features := make([]SnapdFeature, int(lastFeature))
        for i := range features {
                features[i] = SnapdFeature(i)
@@ -91,6 +90,11 @@ func KnownFeatures() []SnapdFeature {
        return features
 }
 
+// KnownFeatures returns the list of all known features.
+func KnownFeatures() []SnapdFeature {
+       return knownFeaturesImpl()
+}
+
 // featureNames maps feature constant to stable string representation.
 // The constants here must be synchronized with cmd/libsnap-confine-private/feature.c
 var featureNames = map[SnapdFeature]string{
```

Someone like @bboozzoo would know what's really best practice in situations like this.

Reply:
That looks much better than what I proposed. I'll change it. Thanks!